mods/default/functions.lua

```diff
@@ -293,12 +293,25 @@ minetest.register_abm({
 -- Dig upwards
 --
 
-function default.dig_up(pos, node, digger)
+local in_dig_up = false
+function default.dig_up(pos, node, digger, max_height)
+	if in_dig_up then return end -- Do not recurse
 	if digger == nil then return end
-	local np = {x = pos.x, y = pos.y + 1, z = pos.z}
-	local nn = minetest.get_node(np)
-	if nn.name == node.name then
-		minetest.node_dig(np, nn, digger)
+	max_height = max_height or 100
+
+	for y = pos.y + 1, pos.y + max_height do
+		local up_pos  = vector.new(pos.x, y, pos.z)
+		local up_node = minetest.get_node(up_pos)
+		if up_node.name == node.name then
+			in_dig_up = true
+			if not minetest.dig_node(up_pos, digger) then
+				in_dig_up = false
+				break
+			end
+			in_dig_up = false
+		else
+			break
+		end
 	end
```

Comment:
Technically, it's not necessary to change the variable more than once:
```suggestion
	in_dig_up = true
	for y = pos.y + 1, pos.y + max_height do
		local up_pos  = vector.new(pos.x, y, pos.z)
		local up_node = minetest.get_node(up_pos)
		if up_node.name ~= node.name then
			break
		end
		if not minetest.dig_node(up_pos, digger) then
			break
		end
	end
	in_dig_up = false
```

Reply:
I decided to duplicate my code because I want `in_dig_up` to be closer to the actual function, so if we are `pcall()`ed (weird error always happens), we are more likely not stuck. Sadly we don't have `finally` keywords like in Python.

Reply:
That sounds like you should use `pcall` (or preferably `xpcall` so you don't lose the stack trace) yourself, then reset the variable after that.

It should be possible to *guarantee* that you aren't stuck. Well, maybe modulo coroutines at least.

ps. I'm not very happy with this mutable global state variable, but I see no good alternative yet.

Reply:
Fixed in 5ae2b04d729254bd33a3c9693a9c389f72652ba2 and f19fa6eb4710f25757098d6f8accdaac35eec5e1